### ISSUE TITLE
Cached run performance with many glob outputs

### ISSUE THREAD

**Issue by valerybugakov** [nrwl/nx#11176]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

It's awesome that [we can use globs in the `outputs` field](https://github.com/nrwl/nx/issues/8593), but when many files are cached, command-runs reusing cache are slow. 

It might be related to [sequential cache checks](https://github.com/nrwl/nx/blob/bd9b33eaef0393d01f747ea9a2ac5d2ca1fb87c6/packages/nx/src/tasks-runner/cache.ts#L343-L347), meaning that the time required to restore the cache grows linearly with the number of cached artifacts. Testing it on the command that generates hundreds of output artifacts (generate CSS module type definitions using `tsm`) makes the cached version 3x slower than the raw command.

## Motivation

Faster execution of the cached command with many outputs cached by a glob pattern.

## Suggested Implementation

It might be that leaning into parallelization of the cache restoration logic used for glob outputs would help.

**Comment by niudai** [comment]
Yeah, we noticed the same behavior. We have hundreds packages with glob outputs, some of which have more than 100 files  in the output folder, and I noticed that even if NX hits the cache, the number of files in outputs still dramatically slows down the command. When the number of files gets larger, running the raw command will even be faster than running nx command with cache hit! 
I hope this issue will be resolved soon! 
Thanks!
